### ISSUE TITLE
[Research] Add EEG Microstate Temporal Features for Emotion (Transition Probabilities + Coverage)

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#56]
## Papers

**Primary:** Emotion recognition based on microstate analysis from temporal and spatial patterns of electroencephalogram
**Authors:** Various
**Year:** 2024
**Venue:** Frontiers in Neuroscience
**URL:** https://www.frontiersin.org/journals/neuroscience/articles/10.3389/fnins.2024.1355512/full

**Supporting:** Cross-subject emotion EEG signal recognition based on source microstate analysis (Frontiers in Neuroscience, 2023)
**URL:** https://www.frontiersin.org/journals/neuroscience/articles/10.3389/fnins.2023.1288580/full

**Supporting:** EEG Microstates in Social and Affective Neuroscience (Brain Topography, Springer, 2023)

## Key Contribution

Proposes KLGEV-criterion-based microstate analysis that automatically identifies optimal number of microstates for emotional EEG. Achieves: SEED 3-class 70.38% (microstate only), 84.47% (microstate + PSD); DEAP valence 75.49%, arousal 74.33% (microstate only), 78.95%/77.61% (+ PSD). Key insight: microstate *temporal* features (duration, occurrence, coverage, transition probabilities) carry emotion information even without full spatial topography.

Neurophysiological finding: arousal positively correlates with microstate C activity (DMN) and negatively with microstate D (dorsal attention network). Valence correlates positively with microstate B (visual network) and negatively with microstate D.

## How It Applies to NeuralDreamWorkshop

**The spatial microstate approach requires 32+ channels and is NOT feasible for Muse 2's 4 channels.** However, the temporal dynamics concept can be adapted:

Instead of clustering full scalp topographies, we can define "microstates" based on the dominant frequency band at each timepoint across our 4 channels. The transitions between these band-dominant states carry temporal information similar to classical microstates.

For example, a rapid alternation between alpha-dominant and beta-dominant states (high transition rate) may indicate emotional instability or regulation effort, while sustained alpha-dominant states indicate calm.

### Files to Modify/Create
- `ml/processing/spectral_microstates.py` — new module (adapted microstate concept for 4-ch)
- `ml/processing/eeg_processor.py` — add `extract_spectral_microstate_features()`
- `ml/models/emotion_classifier.py` — append microstate features to feature vector

### Adapted Algorithm for 4 Channels
```python
import numpy as np
from scipy.signal import welch

def extract_spectral_microstates(signals: np.ndarray, fs: int = 256, 
                                  epoch_ms: int = 250) -> dict:
    """
    Compute spectral microstate features from 4-channel Muse 2 EEG.
    
    Instead of spatial topography clustering (requires 32+ ch),
    we define microstates by dominant frequency band at each time window.
    
    Microstate classes:
      D = delta dominant (0.5-4 Hz)  → deep relaxation / drowsiness
      T = theta dominant (4-8 Hz)    → meditation / creativity / drowsy
      A = alpha dominant (8-12 Hz)   → relaxed wakefulness
      B = beta dominant (12-30 Hz)   → active thinking / stress
    
    Features extracted: duration, occurrence, coverage, transition matrix.
    """
    n_channels, n_samples = signals.shape
    n_epochs = n_samples // (fs * epoch_ms // 1000)
    epoch_len = fs * epoch_ms // 1000  # samples per mini-epoch
    
    bands = {
        'D': (0.5, 4),   # delta
        'T': (4, 8),     # theta
        'A': (8, 12),    # alpha
        'B': (12, 30),   # beta
    }
    
    # 1. Classify each mini-epoch by dominant band (averaged across channels)
    microstate_sequence = []
    for i in range(n_epochs):
        segment = signals[:, i*epoch_len:(i+1)*epoch_len]
        band_powers = {}
        for band_name, (lo, hi) in bands.items():
            freqs, psd = welch(segment, fs=fs, nperseg=min(epoch_len, fs))
            mask = (freqs >= lo) & (freqs <= hi)
            band_powers[band_name] = np.mean(psd[:, mask])
        
        dominant = max(band_powers, key=band_powers.get)
        microstate_sequence.append(dominant)
    
    # 2. Extract temporal features
    states = list(bands.keys())  # ['D', 'T', 'A', 'B']
    features = {}
    
    for s in states:
        occurrences = sum(1 for x in microstate_sequence if x == s)
        # Duration: average consecutive run length
        runs = _get_runs(microstate_sequence, s)
        features[f'ms_{s}_occurrence'] = occurrences / len(microstate_sequence)
        features[f'ms_{s}_coverage'] = occurrences / len(microstate_sequence)
        features[f'ms_{s}_mean_duration'] = np.mean(runs) if runs else 0
        features[f'ms_{s}_max_duration'] = max(runs) if runs else 0
    
    # 3. Transition probability matrix (4x4)
    trans = np.zeros((4, 4))
    for i in range(len(microstate_sequence) - 1):
        from_idx = states.index(microstate_sequence[i])
        to_idx = states.index(microstate_sequence[i + 1])
        trans[from_idx, to_idx] += 1
    
    # Normalize rows
    row_sums = trans.sum(axis=1, keepdims=True)
    row_sums[row_sums == 0] = 1
    trans_prob = trans / row_sums
    
    # Flatten transition matrix as features (16 values)
    for i, si in enumerate(states):
        for j, sj in enumerate(states):
            features[f'ms_trans_{si}_{sj}'] = trans_prob[i, j]
    
    # 4. Derived features
    # Microstate complexity (Shannon entropy of sequence)
    probs = [features[f'ms_{s}_coverage'] for s in states]
    features['ms_entropy'] = -sum(p * np.log(p + 1e-10) for p in probs)
    
    # Alpha-Beta transition rate (marker of emotion regulation)
    features['ms_AB_transition_rate'] = (
        trans_prob[states.index('A'), states.index('B')] +
        trans_prob[states.index('B'), states.index('A')]
    ) / 2
    
    # Theta stability (sustained theta = deep meditation)
    features['ms_theta_stability'] = features.get('ms_T_max_duration', 0)
    
    return features  # ~36 features total

def _get_runs(sequence, target):
    """Get lengths of consecutive runs of target in sequence."""
    runs = []
    current_run = 0
    for s in sequence:
        if s == target:
            current_run += 1
        else:
            if current_run > 0:
                runs.append(current_run)
            current_run = 0
    if current_run > 0:
        runs.append(current_run)
    return runs
```

### Feature Integration
```python
# In extract_features_multichannel():
base_features = ...  # existing 41 features
ms_features = extract_spectral_microstates(signals, fs)
all_features = {**base_features, **ms_features}
# Total: ~77 features (41 existing + 36 microstate)
```

## Expected Performance Impact

- Original paper: microstates alone achieve 70-75% on DEAP (valence/arousal)
- Our adapted spectral microstates will be weaker (no spatial topography), but temporal transition features should still add information
- Expected: +3-5% accuracy when combined with existing PSD features (additive, not replacing)
- Main value: captures *temporal dynamics* of brain state transitions, which current epoch-averaged features miss entirely
- A rapid D→B→A→B pattern tells a different story than sustained A — even if average band powers are identical

## Complexity: LOW

- Pure signal processing — no model training needed for feature extraction
- Adds to existing feature vector for LightGBM (or future GNN from #51)
- 250ms mini-epochs within 4-sec windows → 16 mini-epochs per analysis window
- Computational cost: negligible (<1ms for 4 seconds of data)
- Risk: 4 bands x 4 channels may not have enough diversity for meaningful microstate clustering. Start with channel-averaged approach, test channel-specific if results are promising.

## Dependencies
- None new — uses existing scipy.signal.welch
- Integrates into existing feature extraction pipeline
- Benefits from longer epochs (4-sec windows from issue #33, already implemented)
